### DISCUSSION TITLE
How do I retrieve agent level data from datacollector?

### DISCUSSION THREAD

**Discussion by ghost** [mesa/mesa#2093] [Q&A] [answered]
Hi,

I am able to collect model level data using datacollector. However, what should I do to collect agent level data? Below is my code snip which shows how I am collecting model metrics in the datacollector:
![image](https://github.com/projectmesa/mesa/assets/131771458/e73c2f40-95fe-482c-9e31-920b3a0736e3)
However, I am not sure how to define agent level attributes that I need. I tried something like this:
![image](https://github.com/projectmesa/mesa/assets/131771458/0485b341-76f9-40d5-b47e-1e65d955a8d8)
But it does not work. I am assuming I need to define what attributes of agents I want to collect.

**Comment by EwoutH** [comment]
Hi @PreetamKulkarni! Thanks for reaching out. To collect agent level data, you can define `agent_reporters` in the DataCollector. See:

- [intro_tutorial#collecting-data](https://mesa.readthedocs.io/en/latest/tutorials/intro_tutorial.html#collecting-data)
- [Mesa Data Collection Module](https://mesa.readthedocs.io/en/stable/apis/datacollection.html)

Hope this helps, if you have further questions let us know!

**Reply by ghost** [reply]
Hi @EwoutH,

Thank you for the information. 

Can you please let me know how I can handle multiple agent data? i.e., I have two types of agent and they have some common attributes and some unique. 

Should I write all attributes in "agent_reporters={}"? Is there a way to write multiple agent reporters so that I collect data for all types of agents?

**Reply by EwoutH** [reply, marked as answer]
Unfortunately, this is currently a bit cumbersome. We’re currently discussing on how to enable and improve this, see #348.

The workaround for now is just collect everything for all agents, which will return `None` if they don’t have that attribute.

**Reply by ghost** [reply]
I tried this:
![image](https://github.com/projectmesa/mesa/assets/131771458/db28188c-0c5e-47b8-99ed-641b000415bb)
When I try to run the model, after like 20 seconds, it says"maximum recursion depth exceeded while calling a Python object".
I have 500 agents, is that an issue?

**Reply by ghost** [reply]
When I remove "Agent Type": type(self), it works fine. 

I think self here is not referring to the agent instance.
